docs/doxygen/man/man3/GetSignatures.3
.TH "GetSignatures" 3 "Thu Aug 30 2018" "esc" \" -*- nroff -*-
.ad l
.nh
.SH NAME
GetSignatures \- Class responsible for handling 'get_signatures' command\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <getsignatures\&.h>\fP
.PP
Inherits \fBBlockCommand\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBGetSignatures\fP (uint16_t abank, uint32_t auser, uint32_t ttime, uint32_t block)"
.br
.ti -1c
.RI "virtual int \fBgetType\fP () override"
.br
.RI "\fIReturn TXSTYPE_SIG as type \&. \fP"
.ti -1c
.RI "virtual CommandType \fBgetCommandType\fP () override"
.br
.RI "\fIReturn eReadingOnly as command type \&. \fP"
.ti -1c
.RI "virtual unsigned char * \fBgetData\fP () override"
.br
.RI "\fIGet pointer to command data structure\&. \fP"
.ti -1c
.RI "virtual unsigned char * \fBgetResponse\fP () override"
.br
.RI "\fIGet pointer to response data\&. \fP"
.ti -1c
.RI "virtual void \fBsetData\fP (char *data) override"
.br
.RI "\fIPut data as a char list and put convert it to data structure\&. \fP"
.ti -1c
.RI "virtual void \fBsetResponse\fP (char *response) override"
.br
.RI "\fIApply data as a response struct\&. \fP"
.ti -1c
.RI "virtual int \fBgetDataSize\fP () override"
.br
.RI "\fIGet data struct size\&. Without signature\&. \fP"
.ti -1c
.RI "virtual int \fBgetResponseSize\fP () override"
.br
.RI "\fIGet response data struct size\&. \fP"
.ti -1c
.RI "virtual unsigned char * \fBgetSignature\fP () override"
.br
.RI "\fIGet pointer to signature data\&. \fP"
.ti -1c
.RI "virtual int \fBgetSignatureSize\fP () override"
.br
.RI "\fIGet signature size\&. \fP"
.ti -1c
.RI "virtual void \fBsign\fP (const uint8_t *hash, const uint8_t *sk, const uint8_t *pk) override"
.br
.RI "\fISign actual data plus hash using user private and public keys\&. \fP"
.ti -1c
.RI "virtual bool \fBcheckSignature\fP (const uint8_t *hash, const uint8_t *pk) override"
.br
.RI "\fICheck actual signature\&. \fP"
.ti -1c
.RI "virtual uint32_t \fBgetTime\fP () override"
.br
.RI "\fIGet time of command\&. \fP"
.ti -1c
.RI "virtual uint32_t \fBgetUserId\fP () override"
.br
.RI "\fIGet User ID\&. \fP"
.ti -1c
.RI "virtual uint32_t \fBgetBankId\fP () override"
.br
.RI "\fIGet Node ID\&. \fP"
.ti -1c
.RI "virtual int64_t \fBgetFee\fP () override"
.br
.RI "\fIGet command fee\&. \fP"
.ti -1c
.RI "virtual int64_t \fBgetDeduct\fP () override"
.br
.RI "\fIGet change in cash balance after command\&. \fP"
.ti -1c
.RI "virtual uint32_t \fBgetUserMessageId\fP () override"
.br
.RI "\fIGet message id\&. \fP"
.ti -1c
.RI "virtual bool \fBsend\fP (\fBINetworkClient\fP &netClient) override"
.br
.RI "\fISend data to the server\&. \fP"
.ti -1c
.RI "virtual void \fBsaveResponse\fP (\fBsettings\fP &sts) override"
.br
.RI "\fISave command response to settings object\&. \fP"
.ti -1c
.RI "virtual std::string \fBtoString\fP (bool pretty) override"
.br
.ti -1c
.RI "virtual void \fBtoJson\fP (boost::property_tree::ptree &ptree) override"
.br
.ti -1c
.RI "virtual void \fBtxnToJson\fP (boost::property_tree::ptree &ptree) override"
.br
.ti -1c
.RI "virtual std::string \fBusageHelperToString\fP () override"
.br
.ti -1c
.RI "uint32_t \fBgetBlockNumber\fP ()"
.br
.in -1c
.SS "Additional Inherited Members"
.SH "Detailed Description"
.PP 
Class responsible for handling 'get_signatures' command\&. 
.PP
Definition at line 11 of file getsignatures\&.h\&.
.SH "Member Function Documentation"
.PP 
.SS "bool GetSignatures::checkSignature (const uint8_t * hash, const uint8_t * pk)\fC [override]\fP, \fC [virtual]\fP"

.PP
Check actual signature\&. 
.PP
\fBParameters:\fP
.RS 4
\fIhash\fP Previous hash operation\&. 
.br
\fIpk\fP Pointer to public key\&. 
.RE
.PP

.PP
Implements \fBICommand\fP\&.
.PP
Definition at line 60 of file getsignatures\&.cpp\&.
.SS "bool GetSignatures::send (\fBINetworkClient\fP & netClient)\fC [override]\fP, \fC [virtual]\fP"

.PP
Send data to the server\&. 
.PP
\fBParameters:\fP
.RS 4
\fInetClient\fP Netwrok client implementation of \fBINetworkClient\fP interface\&. 
.br
\fIpk\fP Pointer to public key\&. 
.RE
.PP

.PP
Implements \fBICommand\fP\&.
.PP
Definition at line 88 of file getsignatures\&.cpp\&.
.SS "void GetSignatures::sign (const uint8_t * hash, const uint8_t * sk, const uint8_t * pk)\fC [override]\fP, \fC [virtual]\fP"

.PP
Sign actual data plus hash using user private and public keys\&. 
.PP
\fBParameters:\fP
.RS 4
\fIhash\fP Previous hash operation\&. 
.br
\fIsk\fP Pointer to provate key\&. 
.br
\fIpk\fP Pointer to public key\&. 
.RE
.PP

.PP
Implements \fBICommand\fP\&.
.PP
Definition at line 56 of file getsignatures\&.cpp\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for esc from the source code\&.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.0] - 2018-08-24
### Changed
- Genesis file for the launch
- Update README

## [0.0.6] - 2018-08-10
### Added
- Profit sharing among TOP and VIP nodes

### Changed
- Genesis file for test launch

### Fixed
- Vote counting in get_block command
- Ability to send externally signed transactions

## [0.0.5] - 2018-08-01
### Added
- Support passing version by cmake command.
- Contributing guide
- "Dry run" test case

### Fixed
- Node stops after incorrect `send_many` request  - #134
- Fee checking
- Internal transactions numbers

## [0.0.4] - 2018-07-27
### Changed
- Update default parameters
- Creating a dev version with reference to the last tag

[Unreleased]: https://github.com/adshares/ads/compare/v1.0.0...HEAD

[0.0.5]: https://github.com/adshares/ads/compare/v0.0.6...v1.0.0
[0.0.5]: https://github.com/adshares/ads/compare/v0.0.5...v0.0.6
[0.0.5]: https://github.com/adshares/ads/compare/v0.0.4...v0.0.5
[0.0.4]: https://github.com/adshares/ads/compare/v0.0.3...v0.0.4


src/common/command/getsignatures.cpp
#include "getsignatures.h"
#include "helper/hash.h"
#include "helper/json.h"
#include "helper/signatures.h"

GetSignatures::GetSignatures()
    : m_data{} {
    m_responseError = ErrorCodes::Code::eNone;
}

GetSignatures::GetSignatures(uint16_t abank, uint32_t auser, uint32_t ttime, uint32_t block)
    : m_data{abank, auser, ttime, block} {
    m_responseError = ErrorCodes::Code::eNone;
}

int GetSignatures::getType() {
    return TXSTYPE_SIG;
}

CommandType GetSignatures::getCommandType() {
    return CommandType::eReadingOnly;
}

unsigned char* GetSignatures::getData() {
    return reinterpret_cast<unsigned char*>(&m_data.info);
}

unsigned char* GetSignatures::getResponse() {
    return reinterpret_cast<unsigned char*>(&m_response);
}

void GetSignatures::setData(char* data) {
    m_data = *reinterpret_cast<decltype(m_data)*>(data);
}

void GetSignatures::setResponse(char* response) {
    m_response = *reinterpret_cast<decltype(m_response)*>(response);
}

int GetSignatures::getDataSize() {
    return sizeof(m_data.info);
}

int GetSignatures::getResponseSize() {
    return sizeof(m_response);
}

unsigned char* GetSignatures::getSignature() {
    return m_data.sign;
}

int GetSignatures::getSignatureSize() {
    return sizeof(m_data.sign);
}

void GetSignatures::sign(const uint8_t* /*hash*/, const uint8_t* sk, const uint8_t* pk) {
    ed25519_sign(getData(), getDataSize(), sk, pk, getSignature());
}

bool GetSignatures::checkSignature(const uint8_t* /*hash*/, const uint8_t* pk) {
    return (ed25519_sign_open(getData(), getDataSize(), pk, getSignature()) == 0);
}

uint32_t GetSignatures::getTime() {
    return m_data.info.ttime;
}

uint32_t GetSignatures::getUserId() {
    return m_data.info.auser;
}

uint32_t GetSignatures::getBankId() {
    return m_data.info.abank;
}

int64_t GetSignatures::getFee() {
    return 0;
}

int64_t GetSignatures::getDeduct() {
    return 0;
}

uint32_t GetSignatures::getUserMessageId() {
    return 0;
}

bool GetSignatures::send(INetworkClient& netClient)
{
    if(!netClient.sendData(getData(), sizeof(m_data))) {
        ELOG("GetSignatures sending error\n");
        return false;
    }

    if(!netClient.readData((int32_t*)&m_responseError, ERROR_CODE_LENGTH)) {
        ELOG("GetSignatures reading error\n");
        return false;
    }

    if(m_responseError) {
        return true;
    }

    int32_t signaturesOkSize=0, signaturesNoSize=0;

    if(!netClient.readData((int32_t*)&signaturesOkSize, sizeof(int32_t))) {
        ELOG("GetSignatures reading error\n");
        return false;
    }
    for(int i=0; i<signaturesOkSize; ++i) {
        Helper::Signature sign;
        if(!netClient.readData((char*)&sign, sizeof(sign))) {
            ELOG("GetSignatures reading error\n");
            return false;
        }
        m_signaturesOk.push_back(sign);
    }

    if(!netClient.readData((int32_t*)&signaturesNoSize, sizeof(int32_t))) {
        ELOG("GetSignatures reading error\n");
        return false;
    }
    for(int i=0; i<signaturesNoSize; ++i) {
        Helper::Signature sign;
        if(!netClient.readData((char*)&sign, sizeof(sign))) {
            ELOG("GetSignatures reading error\n");
            return false;
        }
        m_signaturesNo.push_back(sign);
    }

    return true;
}

void GetSignatures::saveResponse(settings&/* sts*/) {
}

std::string GetSignatures::toString(bool /*pretty*/) {
    return "";
}

boost::property_tree::ptree prepareTree(const std::vector<Helper::Signature>& signatures) {
    char sigh[129];
    sigh[128]='\0';

    boost::property_tree::ptree tree;
    for(const auto& sig : signatures) {
        ed25519_key2text(sigh, &sig.signature[0], 64);
        boost::property_tree::ptree sigTree;
        sigTree.put("node", sig.svid);
        sigTree.put("signature", sigh);
        tree.push_back(std::make_pair("", sigTree));
    }
    return tree;
}

void GetSignatures::toJson(boost::property_tree::ptree &ptree) {
    if(!m_responseError) {

        char blockhex[9];
        blockhex[8]='\0';
        sprintf(blockhex,"%08X", getBlockNumber());
        ptree.put("block_time_hex",blockhex);
        ptree.put("block_time", getBlockNumber());

        if(m_signaturesOk.size() > 0) {
            ptree.add_child("signatures", prepareTree(m_signaturesOk));
        }

        if(m_signaturesNo.size() > 0) {
            ptree.add_child("fork_signatures", prepareTree(m_signaturesNo));
        }
    }
    else {
        ptree.put(ERROR_TAG, ErrorCodes().getErrorMsg(m_responseError));
    }
}

void GetSignatures::txnToJson(boost::property_tree::ptree& ptree) {
    using namespace Helper;
    ptree.put(TAG::TYPE, getTxnName(m_data.info.ttype));
    ptree.put(TAG::SRC_NODE, m_data.info.abank);
    ptree.put(TAG::SRC_USER, m_data.info.auser);
    ptree.put(TAG::TIME, m_data.info.ttime);
    ptree.put(TAG::BLOCK, m_data.info.block);
    ptree.put(TAG::SIGN, ed25519_key2text(getSignature(), getSignatureSize()));
}

uint32_t GetSignatures::getBlockNumber() {
    return m_data.info.block;
}

std::string GetSignatures::usageHelperToString() {
    std::stringstream ss{};
    ss << "Usage: " << "{\"run\":\"get_signatures\",[\"block\":<block time as hex>]}" << "\n";
    ss << "Example: " << "{\"run\":\"get_signatures\",\"block\":\"5B1A48C0\"}" << "\n";
    return ss.str();
}
